docs/HighLevelDesignForGraphProgramExtraction.md

```diff
@@ -0,0 +1,70 @@
+
+## Tackling Point 2: Communication Between Swift and TF
+
+Communication between Swift and TF can be handled in a clean manner, by following the same principles.
+
+1. ___Swift -> TF:___ The TF program deals with `Tensor` objects. Therefore, whenever a `Tensor` object is constructed on the Swift side, and later used by TF ops that require tensors, a corresponding `SendToTF` op is added to the graph. This actually consists of two ops: (i) an `Enqueue` op, and a (ii) `Dequeue` op to a size 1 queue. The `Dequeue` op will be used as the `Tensor` object representation for TF ops consuming it later on, and a `session.run(...)` call will be added to the Swift side, enqueueing the constructed tensor that will be sent to TF.
+2. ___TF -> Swift:___ `Tensor` objects allow for various ways to obtain their content, or a reference to them. One proposal could be to have methods like `.scalar`, `.toArray`, `.iterator`, and `.handle`, that return Swift-typed values. Whenever these are used, data needs to be transferred from TF to Swift and so we would add a `ReceiveFromTF` op to the graph. This actually again consists of two ops: (i) an `Enqueue` op, and a (ii) `Dequeue` op to a size 1 queue. The `Enqueue` op will be added to the TF graph, enqueueing the corresponding TF tensor (or its handle, for example) to the queue. At the same time, a `session.run(...)` call will be added to the Swift side, dequeueing the TF content that is to be received from the graph. These method calls will be required to be made explicitly from the library user and will not be added automatically by the Swift compiler, in order to avoid confusion and potential inefficiencies.
+
+## Tacking Point 3: Handling Functions that use TF
+
+Currently, Swift for TF handles all Swift functions that use TF by aggressively inlining them. This can cause lots of issues, including an exploding size of the compiled code, as well as limited supported for recursion (note that TF graph functions themselves support recursion and thus we should be able to support it on our side too). Here we suggest we use TF graph functions to handle Swift functions that use TF. We propose the following high-level design: all Swift functions that use TF will be compiled into two separate functions:
+
+1. ___Swift Function:___ The Swift code part of the compiled function that contains all computation that needs to happen on the Swift side. Note that in the case that this is a pure TF function (using only TF ops), then this function could be compiled to a no-op, or optimized away entirely.
+2. ___TF Graph Function:___ The TF part of the compiled function that is a TF `FunctionDef` protobuf message which contains the extracted TF graph for the compiled function.
```

Comment:
You should mention here what do to about the function signatures, and which information needs to be made available to swift and to tensorflow.

I think we can compute the swift-side and tf-side function signatures and store them somewhere as part of the type of this composite function. This needs to be visible at compile time so the compiler can build calls to tf functions without seeing its definition.

At link time (or runtime initialization, whatever is more convenient), when all the definitions are available, the full graph can be stitched together.

Reply:
Richard's comment further down is relevant for this. I'll respond there.